File: src/services.rs
use crate::utils::{RandTable, Tcopy};
use std::fmt;

#[derive(Clone)]
pub struct Service {
    pub name: &'static str,
    pub wait_time: isize,
    pub cost: isize,
    pub demand: isize,
    pub profit: isize,
    pub offer: isize,
    pub usage: isize,
    pub backends: Option<Vec<Service>>,
}

impl Service {
    pub fn new(
        name: &'static str,
        wait_time: isize,
        cost: isize,
        demand: isize,
        profit: isize,
        offer: isize,
        usage: isize,
        backends: Option<Vec<Service>>,
    ) -> Self {
        Service {
            name,
            wait_time,
            cost,
            demand,
            profit,
            offer,
            usage,
            backends,
        }
    }

    pub fn set_config(&mut self, property: &str, value: isize) -> Result<(), &'static str> {
        match property {
            "name" => return Err("Can't change `name` value"),
            "wait_time" => self.wait_time = value,
            "cost" => self.cost = value,
            "demand" => self.demand = value,
            "profit" => self.profit = value,
            "offer" => self.offer = value,
            "usage" => self.usage = value,
            "backends" => return Err("Use `set_backend_config()` to configure backends"),
            _ => return Err("No such property name"),
        }

        Ok(())
    }

    pub fn add_backend(
        &mut self,
        name: &'static str,
        wait_time: isize,
        cost: isize,
        demand: isize,
        profit: isize,
        offer: isize,
        usage: isize,
        backends: Option<Vec<Service>>,
    ) {
        let backend = Service {
            name,
            wait_time,
            cost,
            demand,
            profit,
            offer,
            usage,
            backends,
        };

        self.backends
            .get_or_insert_with(|| Vec::new())
            .push(backend);
    }

    pub fn set_backend_config(
        backend: &mut Service,
        property: &str,
        value: isize,
    ) -> Result<(), &'static str> {
        Ok(backend.set_config(property, value)?)
    }
}

impl Tcopy for Service {
    fn copy(&mut self) -> Service {
        Service {
            name: self.name,
            wait_time: self.wait_time,
            cost: self.cost,
            demand: self.demand,
            profit: self.profit,
            offer: self.offer,
            usage: self.usage,
            backends: Some((*self.backends.get_or_insert_with(|| Vec::new()).clone()).to_vec()),
        }
    }
}

impl fmt::Display for Service {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}", self.name)
    }
}

pub struct ServiceBucket {
    pub services: Vec<Service>,
    pub weights: Vec<u8>,
}

impl ServiceBucket {
    pub fn new() -> Self {
        Self {
            services: Vec::new(),
            weights: Vec::new(),
        }
    }

    pub fn push(&mut self, service: Service, weight: u8) {
        self.services.push(service);
        self.weights.push(weight);
    }

    pub fn find(&mut self, name: &'static str) -> Option<Service> {
        for service in &self.services {
            if service.name == name {
                return Some(service.clone());
            }
        }

        None
    }

    pub fn random(&mut self) -> Service {
        //
        // Need to see if it'd be worth only calculating the random table after each insert instead
        // of every time the random function is called.
        //
        RandTable::new(self.services.clone(), self.weights.clone()).random()
    }
}

File: c++/services.cpp
//
// Created by Tommy Pujol on 4/19/21.
//

#include "services.h"

void ServiceBucket::push(Service& service, int weight) {
    this->services.push_back(service);
    this->weights.push_back(weight);
    this->renew_random_table();
}

[[maybe_unused]]
std::optional<Service> ServiceBucket::find(const char* name) {
    // FIXME: Use a more efficient search algorithm.
    for (auto& service : this->services) {
        if(strcmp(service.name, name) == 0) return service;
    }

    return {};
}

Service* ServiceBucket::random() {
    return this->services_table.random();
}

[[maybe_unused]]
void ServiceBucket::remove(const char *name) const {
    vector<Service> vec = this->services;
    int idx = 0;
    for (auto& service : vec) {
        if (service.name == name) {
            break;
        }

        idx++;
    }

    vec.erase(vec.begin() + idx);
}

void ServiceBucket::renew_random_table() {
    this->services_table = RandTable<Service>(this->services, this->weights);
}

[[maybe_unused]]
int Service::set_config(const char *property, long new_value) {

    if (strcmp(property, "wait_time") == 0) { this->wait_time = new_value; return 0; }
    else if (strcmp(property, "cost") == 0) { this->cost = new_value; return 0; }
    else if (strcmp(property, "demand") == 0) { this->demand = new_value; return 0; }
    else if (strcmp(property, "profit") == 0) { this->profit = new_value; return 0; }
    else if (strcmp(property, "offer") == 0) { this->offer = new_value; return 0; }
    else if (strcmp(property, "usage") == 0) { this->usage = new_value; return 0; }
    else { return -1; }
}

[[maybe_unused]]
void Service::add_backend(const Service& backend) {
    this->backends.push_back(backend);
}